torchtune/modules/model_fusion/fusion_embed.py

```diff
@@ -0,0 +1,74 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List
+
+import torch
+from torch import nn, Tensor
+
+
+class FusionEmbedding(nn.Module):
+    """Fusion embedding supports training additional special tokens while keeping
+    the original embedding frozen. When fusing new models with a langauge model,
+    there may be some additional tokens needed to support the fused langauge model.
+    The FusionEmbedding keeps the original embeddings frozen while learning a much smaller
+    second embedding for the additional tokens. During forward this module routes
+    the tokens to the appropriate embedding table.
+
+    Args:
+        vocab_size (int): language model vocab size
+        additional_tokens (int): additional tokens for the fused model
+        embed_dim (int): embedding dimension of the two embedding tables
+    """
+
+    def __init__(self, vocab_size: int, additional_tokens: int, embed_dim: int) -> None:
+        super().__init__()
+        self.embedding = nn.Embedding(vocab_size, embed_dim)
+        self.fusion_embedding = nn.Embedding(additional_tokens, embed_dim)
+        self.dim = embed_dim
+        # TODO: Support merging the embeddings after finetuning
+
+    def fusion_params(self) -> List[str]:
+        """
+        Return fusion embedding parameters.
+        """
+        fusion_params = ["fusion_embedding.weight"]
+        return fusion_params
+
+    def _fused_embed(self, bs, seq_len):
+        """
+        Return an empty tensor the shape of the combined embedding.
+        """
+        device = self.embedding.weight.device
+        dtype = self.embedding.weight.dtype
+        return torch.empty(bs, seq_len, self.dim, device=device, dtype=dtype)
+
+    def forward(self, input: Tensor) -> Tensor:
+        """
+        Args:
+            input (Tensor): input integer tensor with shape
+                [batch_size x seq_length]
+
+        Returns:
+            Tensor: output tensor embedding with shape
+                [batch_size x seq_length x embed_dim]`
+
+        """
+        bs, seq_len = input.size()
+        vocab_size = self.embedding.num_embeddings
```

Comment:
nit: could just save vocab_size as attribute directly in init?

Reply:
Each embedding keeps its own size already, it feels redundant for us to as well. Though I should probably add the property "num_embeddings" so this could be used like an embedding.